be/src/exec/vectorized/hdfs_scanner_text.cpp

```diff
@@ -0,0 +1,190 @@
+
+#include "exec/vectorized/hdfs_scan_node.h"
+#include "gutil/strings/substitute.h"
+#include "util/utf8_check.h"
+
+namespace starrocks::vectorized {
+
+Status HdfsTextScanner::CSVReader::next_record(Record* record) {
+    if (_limit > 0 && _parsed_bytes > _limit) {
+        return Status::EndOfFile("Reached limit");
+    }
+    char* d;
+    size_t pos = 0;
+    while ((d = _buff.find(_record_delimiter, pos)) == nullptr) {
+        pos = _buff.available();
+        _buff.compact();
+        if (_buff.free_space() == 0) {
+            RETURN_IF_ERROR(_expand_buffer());
+        }
+        RETURN_IF_ERROR(_fill_buffer());
+    }
+    size_t l = d - _buff.position();
+    *record = Record(_buff.position(), l);
+    _buff.skip(l + 1);
+    //               ^^ skip record delimiter.
+    _parsed_bytes += l + 1;
+    return Status::OK();
+}
+
+Status HdfsTextScanner::CSVReader::_fill_buffer() {
+    DCHECK(_buff.free_space() > 0);
+    Slice s(_buff.limit(), _buff.free_space());
+    Status st = _file->read(_offset, &s);
+    _offset += s.size;
+    // According to the specification of `Env::read`, when reached the end of
+    // a file, the returned status will be OK instead of EOF, but here we check
+    // EOF also for safety.
+    if (st.is_end_of_file()) {
+        s.size = 0;
+    } else if (!st.ok()) {
+        return st;
+    }
+    _buff.add_limit(s.size);
+    auto n = _buff.available();
+    if (s.size == 0 && n == 0) {
+        // Has reached the end of file and the buffer is empty.
+        return Status::EndOfFile(_file->file_name());
+    } else if (s.size == 0 && _buff.position()[n - 1] != _record_delimiter) {
+        // Has reached the end of file but still no record delimiter found, which
+        // is valid, according the RFC, add the record delimiter ourself.
+        _buff.append(_record_delimiter);
+    }
+    return Status::OK();
+}
+
+Status HdfsTextScanner::CSVReader::_expand_buffer() {
+    if (UNLIKELY(_storage.size() >= kMaxBufferSize)) {
+        return Status::InternalError("CSV line length exceed limit " + std::to_string(kMaxBufferSize));
+    }
+    size_t new_capacity = std::min(_storage.size() * 2, kMaxBufferSize);
+    DCHECK_EQ(_storage.data(), _buff.position()) << "should compact buffer before expand";
+    _storage.resize(new_capacity);
+    Buffer new_buff(_storage.data(), _storage.size());
+    new_buff.add_limit(_buff.available());
+    DCHECK_EQ(_storage.data(), new_buff.position());
+    DCHECK_EQ(_buff.available(), new_buff.available());
+    _buff = new_buff;
+    return Status::OK();
+}
+
+void HdfsTextScanner::CSVReader::split_record(const Record& record, Fields* fields) const {
+    const char* value = record.data;
+    const char* ptr = record.data;
+    const size_t size = record.size;
+
+    if (_field_delimiter.size() == 1) {
+        for (size_t i = 0; i < size; ++i, ++ptr) {
+            if (*ptr == _field_delimiter[0]) {
```

Comment:
Do we need to consider escape or quote char? For example

if  `field_delimieter` is "," and csv line is "hello,"world,yes",ok", records should be splitted to ["hello", "world,yes", "ok"] ?

Reply:
I want to submit an other pr for this since csv scanner also depend on this @dirtysalt